programs/exchange/src/math.rs

```diff
@@ -554,27 +683,81 @@ mod tests {
                 price: 2 * 10u64.pow(PRICE_OFFSET.into()),
                 ..Default::default()
             };
-            let result = calculate_max_withdrawable(&asset, 100 * 10u64.pow(6));
-            assert_eq!(result, 50 * 10u64.pow(6))
+            let max_withdrawable = calculate_max_withdrawable(&asset, 0u64);
+            assert_eq!(max_withdrawable, 0u64);
+        }
+        {
+            let asset = Asset {
+                decimals: 6,
+                price: 2 * 10u64.pow(PRICE_OFFSET.into()),
+                ..Default::default()
+            };
+            let max_withdrawable = calculate_max_withdrawable(&asset, 100 * 10u64.pow(6));
+            assert_eq!(max_withdrawable, 50 * 10u64.pow(6))
         }
     }
     #[test]
     fn test_amount_to_shares() {
+        // not initialized shares
         {
-            let result = amount_to_shares(10, 100, 10);
-            assert_eq!(result, 1)
+            let amount = amount_to_shares(0, 0, 0);
+            assert_eq!(amount, 0)
+        }
+        // zero amount
+        {
+            let amount = amount_to_shares(100, 100 * 10u64.pow(6), 0);
+            assert_eq!(amount, 0)
+        }
+        // basic
+        {
+            let amount = amount_to_shares(10, 100, 10);
+            // 1/10 of all_shares
+            assert_eq!(amount, 1)
+        }
+        // large numbers
+        {
+            let amount = amount_to_shares(
+                10u64.pow(6),
```

Comment:
I wouldn't consider it 'large' since each asset will have at least 6 decimals.
Large i would consider 10 ** 18